File: 20230306_Monday/20230306.go
// Leetcode: 1539. Kth Missing Positive Number.
// https://leetcode.com/problems/kth-missing-positive-number/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Kth Missing Positive Number.
// Memory Usage: 2.9 MB, less than 18.18% of Go online submissions for Kth Missing Positive Number.
// 2023.03.06 Daily Challenge.
func findKthPositive(arr []int, k int) int {
  lst,cnt := 0, 0 // Last and counter variables.
  for _,x := range arr {
    msd := x - lst - 1 // Amount of missed elements.
    // Explore new gap between two elements.
    if 0 != msd {
      ncnt := cnt + msd // New counter.
      // k-th element is in this gap.
      if k <= ncnt {
        return lst + (k - cnt)
      }
      cnt = ncnt
    }
    lst = x
  }
  return lst  + (k - cnt)
}

File: 20230306_Monday/20230306.rb
# Leetcode: 1539. Kth Missing Positive Number.
# https://leetcode.com/problems/kth-missing-positive-number/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 75 ms, faster than 86.36% of Ruby online submissions for Kth Missing Positive Number.
# Memory Usage: 211.3 MB, less than 77.27% of Ruby online submissions for Kth Missing Positive Number.
# @param {Integer[]} arr
# @param {Integer} k
# @return {Integer}
def find_kth_positive(arr, k)
  # Caluclating gaps between last unmissed and current element.
  
  # Save last number that is not missed.
  # If starting from 1 it coung to be 0.
  last = 0
  # Counter of missed number.
  counter = 0
  # Iterate over elements in array.
  # Calculating amount of vissed number by: curr - last - 1.
  # If there are NO missed numbers keep on moving.
  # If there are - implement some math.
  arr.each do |x|
    missed = x - last - 1 # Amount of missed elements.
    if 0 != missed
      new_counter = counter + missed
      if k <= new_counter
        diff = k - counter
        return last + diff
      end
      counter = new_counter
    end
    last = x
  end
  return last + (k - counter)
end
